File: src/adapter/php/adapter.php
<?php
/**
 * 对象适配器使用的是委派
 */

interface Target {
	public function request();
}

class Adaptee {
	public function specificRequest() {
		printf("%s\n", __FUNCTION__);
	}
}

class Adapter implements Target {
	private $_adaptee = null;

	public function __construct(Adaptee $adaptee) {
		$this->_adaptee = $adaptee;
	}

	/**
	 * 委派调用Adaptee的request方法
	 */
	public function request() {
		$this->_adaptee->specificRequest();
	}

	public function __destruct() {
		$this->_adaptee = null;
	}
}

class Client {
	public static function main() {
		$adaptee = new Adaptee();
		$adapter = new Adapter($adaptee);
		$adapter->request();
		$adaptee = null;
		$adapter = null;
	}
}

Client::main();

?>

File: src/java/src/main/java/com/errpro/design_patterns/adapter/Adapter.java
package com.errpro.design_patterns.adapter;

// 1)对象适配器
public class Adapter implements Target {
    private Adaptee adaptee = null;

    public Adapter(Adaptee adaptee) {
        this.adaptee = adaptee;
    }

    public void request() {
        adaptee.specificRequest();
    }
}


// 2)类适配器
/*
public class Adapter extends Adaptee implements Target {
	public void request() {
		super.specificRequest();
	}
}
*/
